lib/trino-aws/src/main/java/io/trino/aws/AwsCredentialsProviderModule.java

```diff
@@ -0,0 +1,150 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.aws;
+
+import com.amazonaws.auth.AWSCredentials;
+import com.amazonaws.auth.AWSCredentialsProvider;
+import com.amazonaws.auth.AWSStaticCredentialsProvider;
+import com.amazonaws.auth.BasicAWSCredentials;
+import com.amazonaws.auth.BasicSessionCredentials;
+import com.amazonaws.auth.DefaultAWSCredentialsProviderChain;
+import com.amazonaws.auth.STSAssumeRoleSessionCredentialsProvider;
+import com.amazonaws.services.securitytoken.AWSSecurityTokenServiceClientBuilder;
+import com.google.inject.Binder;
+import com.google.inject.Module;
+import com.google.inject.Provides;
+
+import javax.inject.Singleton;
+
+import java.util.Optional;
+
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * This Guice {@link Module} will bind an {@link AWSCredentialsProvider} in the
+ * Guice context. The provider is configured using an injected configuration,
+ * which must implement the {@link AwsCredentialsProviderConfig} interface. The actual
+ * type of the configuration class is a type parameter on the module class; for
+ * this to work, this type parameter must be "reified" by using a trick with
+ * creating an anonymous subclass of this module:
+ * {@code install(new AwsCredentialsProviderModule<ActualConfig>() {})}. That
+ * way the {@link Provides}-annotated method {@link #getAwsCredentialsProvider(AwsCredentialsProviderConfig)}
+ * will actually contain the instantiation of {@link Config} as its parameter's
+ * type and Guice will be able to resolve it for injection. You may be familiar
+ * with this pattern from using {@link com.google.inject.TypeLiteral}.
```

Comment:
Is it too much magic? Perhaps it would be possible to do it somehow by passing a `Class<Config>` and not do dirty tricks with anonymous classes.